flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,361 @@
+            executionSlotSharingGroup.addVertex(executionVertexId);
+            executionSlotSharingGroupMap.put(executionVertexId, executionSlotSharingGroup);
+        }
+
+        private void updateConstraintToExecutionSlotSharingGroupMap(
+                final List<SchedulingExecutionVertex> executionVertices) {
+
+            for (SchedulingExecutionVertex executionVertex : executionVertices) {
+                final ExecutionVertexID executionVertexId = executionVertex.getId();
+                final CoLocationGroup coLocationGroup =
+                        coLocationGroupMap.get(executionVertexId.getJobVertexId());
+                if (Objects.nonNull(coLocationGroup)) {
+                    final CoLocationConstraint constraint =
+                            getCoLocationConstraint(executionVertex);
+
+                    constraintToExecutionSlotSharingGroupMap.put(
+                            constraint, executionSlotSharingGroupMap.get(executionVertexId));
+                }
+            }
+        }
+
+        private CoLocationConstraint getCoLocationConstraint(SchedulingExecutionVertex sev) {
+            final JobVertexID jobVertexID = sev.getId().getJobVertexId();
+            final int subtaskIndex = sev.getId().getSubtaskIndex();
+            return coLocationGroupMap.get(jobVertexID).getLocationConstraint(subtaskIndex);
+        }
+
+        private int getSlotRoundRobinIndex(
+                final int jobVertexParallelism, SlotSharingGroup slotSharingGroup) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    paralleledExecutionSlotSharingGroupsMap.get(slotSharingGroup);
+            final boolean maxParallel = jobVertexParallelism == executionSlotSharingGroups.size();
+            return maxParallel ? 0 : slotSharingGroupIndexMap.getOrDefault(slotSharingGroup, 0);
+        }
+
+        private int getFittestSlotIndex(
+                final List<ExecutionSlotSharingGroup> executionSlotSharingGroups,
+                final @Nullable SchedulingExecutionVertex executionVertex) {
+            int indexWithLeastExecutionVertices = 0;
+            int leastExecutionVertices = Integer.MAX_VALUE;
+            for (int index = 0; index < executionSlotSharingGroups.size(); index++) {
+                final ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        executionSlotSharingGroups.get(index);
+                final int executionVertices =
+                        executionSlotSharingGroup.getExecutionVertexIds().size();
+                if (leastExecutionVertices > executionVertices
+                        && allocatable(executionSlotSharingGroup, executionVertex)) {
+                    indexWithLeastExecutionVertices = index;
+                    leastExecutionVertices = executionVertices;
+                }
+            }
+            return indexWithLeastExecutionVertices;
+        }
+
+        private void updateSlotRoundRobinIndexIfNeeded(
```

Comment:
The two methods `updateSlotRoundRobinIndexIfNeeded` and `getSlotRoundRobinIndex` are best placed next to each other.

And it's better to extract an method to check whether current paralleslism is the maxParallel of current ssg. These 2 methods will use it. Currently, you write it twice.

Reply:
Also, I found `allocateCoLocatedVertices` also check whether current paralleslism is the maxParallel of current ssg. So 3 callers can call the common method.